File: python/2807.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def gcd(a: int, b: int) -> int:
            if b == 0:
                return a
            return gcd(b, a % b)
        pre, nxt = head, head.next
        while nxt != None:
            x = gcd(pre.val, nxt.val)
            pre.next = ListNode(x, nxt)
            pre = nxt
            nxt = nxt.next
        return head

File: go/2807.go
package main

func insertGreatestCommonDivisors(head *ListNode) *ListNode {
	var gcd func(a, b int) int
	gcd = func(a, b int) int {
		if b == 0 {
			return a
		}
		return gcd(b, a%b)
	}
	pre, nxt := head, head.Next
	for nxt != nil {
		x := gcd(pre.Val, nxt.Val)
		pre.Next = &ListNode{x, nxt}
		pre = nxt
		nxt = nxt.Next
	}
	return head
}
